src/core/localize.rkt

```diff
@@ -191,73 +191,174 @@
-                              'actual-values
-                              (vector->list (vector-ref actuals-out n))))
+      (begin0 (values subexpr (vector->list (vector-ref errs n)))
         (set! n (add1 n))))))
 
+(module+ test
+  (define ctx (make-debug-context '(x y)))
+  (define spec `(- (sqrt (+ x 1)) (sqrt y)))
+  (define pt `(1e-100 1e-100))
+  (define exact 1e-50)
+  (check-equal? (first (apply (eval-progs-real (list `(- ,spec ,exact)) (list ctx)) pt)) 1.0))
+
 ;; Compute the local error of every subexpression of `prog`
 ;; and returns the error information as an S-expr in the
 ;; same shape as `prog`
 (define (local-error-as-tree test ctx)
-  (define errs (first (compute-local-errors (list (all-subexpressions (test-input test))) ctx)))
-
-  (define local-error
-    (let loop ([expr (test-input test)])
-      (define expr-info (hash-ref errs expr))
-      (define err-list (hash-ref expr-info 'errs))
-      (match expr
-        [(list op args ...) (cons err-list (map loop args))]
-        [_ (list err-list)])))
-
-  (define exact-values
-    (let loop ([expr (test-input test)])
-      (define expr-info (hash-ref errs expr))
-      (define exacts-list (hash-ref expr-info 'exact-values))
-      (match expr
-        [(list op args ...) (cons exacts-list (map loop args))]
-        [_ (list exacts-list)])))
-
-  (define actual-values
-    (let loop ([expr (test-input test)])
-      (define expr-info (hash-ref errs expr))
-      (define actual-list (hash-ref expr-info 'actual-values))
-      (match expr
-        [(list op args ...) (cons actual-list (map loop args))]
-        [_ (list actual-list)])))
-
-  (define tree
-    (let loop ([expr (prog->fpcore (test-input test) (test-context test))]
-               [err local-error]
-               [exact exact-values]
-               [actual actual-values])
-      (match expr
-        [(list op args ...)
-         ;; err => (List (listof Integer) List ...)
-         (hasheq 'e
-                 (~a op)
-                 'avg-error
-                 (format-bits (errors-score (first err)))
-                 'exact-value
-                 (map ~s (first exact))
-                 'actual-value
-                 (map ~s (first actual))
-                 'children
-                 (map loop args (rest err) (rest exact) (rest actual)))]
-        ;; err => (List (listof Integer))
-        [_
-         (hasheq 'e
-                 (~a expr)
-                 'avg-error
-                 (format-bits (errors-score (first err)))
-                 'exact-value
-                 (map ~s (first exact))
-                 'actual-value
-                 (map ~s (first actual))
-                 'children
-                 '())])))
-  tree)
+  (define fpcore (prog->fpcore (test-input test) (test-context test)))
+  (define subexprss (list (all-subexpressions (test-input test))))
+  (define exprs-list (append* subexprss)) ; unroll subexprss
```

Comment:
This is fine but I think `(append* (list x))` is just `x`, you don't need to do this.